sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,625 @@
+        return status;
+    }
+
+    nanocbor_value_t pub_key;
+    nanocbor_skip(key); /* Skip to public key */
+    res = nanocbor_enter_array(key, &pub_key);
+    if (res < 0) {
+        goto error;
+    }
+    return psa_decode_key_data(&pub_key, pubkey_data,
+                                    pubkey_data_len);
+error:
+    DEBUG("[persistent storage] error while decoding asymmetric key slot: %s",
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+#endif /* PSA_ASYMMETRIC_KEYPAIR_COUNT */
+
+psa_status_t psa_decode_key_attributes(psa_key_attributes_t *attr,
+                                       uint8_t *cbor_buf, size_t cbor_buf_size)
+{
+    int res = 0;
+
+    nanocbor_value_t dec;
+    nanocbor_decoder_init(&dec, cbor_buf, cbor_buf_size);
+
+    nanocbor_value_t key_slot;
+    res = nanocbor_enter_array(&dec, &key_slot);
+    if (res < 0) {
+        goto error;
+    }
+
+    nanocbor_value_t attr_dec;
+    res = nanocbor_enter_array(&key_slot, &attr_dec);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_get_uint32(&attr_dec, &attr->id);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_get_uint16(&attr_dec, &attr->type);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_get_uint16(&attr_dec, &attr->bits);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_get_uint32(&attr_dec, &attr->lifetime);
+    if (res < 0) {
+        goto error;
+    }
+
+    nanocbor_value_t policy;
+    res = nanocbor_enter_array(&attr_dec, &policy);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_get_uint32(&policy, &attr->policy.usage);
+    if (res < 0) {
+        goto error;
+    }
+    res = nanocbor_get_uint32(&policy, &attr->policy.alg);
+    if (res < 0) {
+        goto error;
+    }
+
+#if ENABLE_DEBUG
+    puts("Key Attributes:\n");
+    printf("ID: %x\n", (int) attr->id);
+    printf("Type: %x\n", (int) attr->type);
+    printf("Bits: %d\n", (int) attr->bits);
+    printf("Lifetime: %x\n", (int) attr->lifetime);
+    printf("Usage: %x\n", (int) attr->policy.usage);
+    printf("Algorithm: %x\n", (int) attr->policy.alg);
+#endif
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while decoding key attributes: %s",
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+
+psa_status_t psa_decode_key_slot_data(psa_key_slot_t *slot, uint8_t *cbor_buf, size_t cbor_buf_size)
+{
+    int res = 0;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    nanocbor_value_t dec;
+    nanocbor_decoder_init(&dec, cbor_buf, cbor_buf_size);
+
+    nanocbor_value_t key_slot;
+    res = nanocbor_enter_array(&dec, &key_slot);
+    if (res < 0) {
+        goto error;
+    }
+
+    /* We should have decoded the attributes before, so we can skip them here */
+    nanocbor_skip(&key_slot);
```

Comment:
Couldn't we initialize the `nanocbor_value_t` in the calling function and reuse it for both decoding attributes and the data?

Reply:
If we do this, the key slot management needs to know about nanocbor. To have a clear separation I'd like to only use nanocbor in the encoder module.

Reply:
Makes sense, I second this!